File: darkmode/weather.py
import requests
import numpy as np
from sklearn.linear_model import LinearRegression
from datetime import datetime
from sklearn.metrics import mean_squared_error, r2_score

# If the key is new, wait up to 2 hours for activation.
API_KEY = "your_openweathermap_api_key_here"

def analyze_weather(city_name):
    # 1. Fetch Forecast Data
    url = f"https://api.openweathermap.org/data/2.5/forecast?q={city_name}&appid={API_KEY}&units=imperial"
    
    try:
        response = requests.get(url, timeout=10)
        # Check for activation (401) or misspelled city (404) errors
        if response.status_code != 200:
            error_msg = response.json().get('message', 'API request failed')
            return {"error": f"API Error: {error_msg} (Status: {response.status_code})"}
    except requests.RequestException as e:
        return {"error": f"Connection error: {str(e)}"}

    data = response.json()
    
    # 2. Extract Data (Next 24 hours = 8 points)
    forecast_slice = data['list'][:8]
    temps = [item['main']['temp'] for item in forecast_slice]
    labels = [datetime.fromtimestamp(item['dt']).strftime('%I:%M %p') for item in forecast_slice]
    
    # 3. Prepare for Scikit-Learn
    X = np.array(range(len(temps))).reshape(-1, 1) 
    y = np.array(temps)

    # 4. Perform Linear Regression
    model = LinearRegression()
    model.fit(X, y)
    
    slope = model.coef_[0]
    intercept = model.intercept_
    
    # Generate Predictions for evaluation
    y_pred = model.predict(X)
    trend_line = y_pred.tolist() 

    # 5. Evaluation Metrics
    # R2 (R-squared) measures fit; closer to 1.0 is better.
    # MSE (Mean Squared Error) measures average error; lower is better.
    r2 = r2_score(y, y_pred)
    mse = mean_squared_error(y, y_pred)

    # 6. Analyze Trend
    trend_text = "Stable"
    if slope > 0.5:
        trend_text = "Heating Up 📈"
    elif slope < -0.5:
        trend_text = "Cooling Down 📉"

    # 7. Extract Details for UI
    current = data['list'][0]
    
    return {
        "city": data['city']['name'],
        "current_temp_f": round(current['main']['temp']),
        "current_temp_c": round((current['main']['temp'] - 32) * 5/9),
        "wind_speed": f"{current['wind']['speed']:.2f} mph",
        "time_of_day": datetime.fromtimestamp(current['dt']).strftime('%I:%M:%S %p'),
        "labels": labels,
        "actual_temps": temps,
        "trend_line": trend_line,
        "trend_text": trend_text,
        "slope": slope,
        "mse": round(mse, 4), 
        "r2": round(r2, 4)
    }

File: darkmode/weather.js
let myChart = null;
let currentUnit = 'F'; // Default unit
let currentData = null; // Store data for unit toggling

// --- DOM ELEMENTS ---
const cityInput = document.getElementById("city-input");
const tempDisplay = document.getElementById("temp-display");
const weatherDesc = document.getElementById("weather-desc"); // You might need to add this ID to HTML if missing, or use existing logic
const feelsLike = document.getElementById("feels-like");
const aiText = document.getElementById("ai-prediction");
const locationText = document.getElementById("current-location");
const timeDisplay = document.getElementById("time-display");

// Stats Elements
const r2Val = document.getElementById("r2-val");
const mseVal = document.getElementById("mse-val");
const windVal = document.getElementById("wind-val");
const slopeVal = document.getElementById("slope-val");

// --- 1. FETCH DATA FROM PYTHON ---
function getWeather(cityName) {
    // Show loading state
    aiText.innerText = "Analyzing...";
    aiText.style.color = "#ccc";

    fetch(`/api/analyze?city=${cityName}`)
        .then(response => {
            if (!response.ok) throw new Error("City not found");
            return response.json();
        })
        .then(data => {
            if(data.error) throw new Error(data.error);
            currentData = data; // Save for unit switching
            updateUI(data);
        })
        .catch(error => {
            console.error(error);
            aiText.innerText = "Error: " + error.message;
            aiText.style.color = "#ff4d4d";
        });
}

// --- 2. UPDATE THE UI ---
function updateUI(data) {
    // A. Update Text Info
    locationText.innerText = data.city;
    timeDisplay.innerText = data.time_of_day;
    windVal.innerText = data.wind_speed;
    
    // Model Stats
    r2Val.innerText = data.r2; 
    mseVal.innerText = data.mse;
    slopeVal.innerText = data.slope.toFixed(4);

    // B. Update Temperature (Handle Units)
    updateTempDisplay();

    // C. Update AI Verdict
    aiText.innerText = data.trend_text;
    if(data.trend_text.includes("Heating")) {
        aiText.style.color = "#ffeb3b"; // Yellow/Hot
    } else if(data.trend_text.includes("Cooling")) {
        aiText.style.color = "#4facfe"; // Blue/Cool
    } else {
        aiText.style.color = "#fff"; // White/Stable
    }

    // D. Update Graph
    updateChart(data);
}

// --- 3. HANDLE UNITS (C/F) ---
function updateTempDisplay() {
    if(!currentData) return;

    if(currentUnit === 'F') {
        tempDisplay.innerText = currentData.current_temp_f + "°";
        // Simple calculation for "feels like" if not provided by API, or just use temp
        feelsLike.innerText = (currentData.current_temp_f - 2) + "°"; 
    } else {
        tempDisplay.innerText = currentData.current_temp_c + "°";
        feelsLike.innerText = (currentData.current_temp_c - 1) + "°";
    }
}

// Event Listeners for Buttons
document.getElementById('btn-c').addEventListener('click', (e) => {
    currentUnit = 'C';
    e.target.classList.add('active');
    document.getElementById('btn-f').classList.remove('active');
    updateTempDisplay();
});

document.getElementById('btn-f').addEventListener('click', (e) => {
    currentUnit = 'F';
    e.target.classList.add('active');
    document.getElementById('btn-c').classList.remove('active');
    updateTempDisplay();
});

// --- 4. RENDER CHART (The Linear Regression Visuals) ---
function updateChart(data) {
    const ctx = document.getElementById('weatherChart').getContext('2d');
    
    if (myChart) myChart.destroy();

    myChart = new Chart(ctx, {
        type: 'line',
        data: {
            labels: data.labels, // From Python
            datasets: [
                {
                    label: 'Actual Forecast',
                    data: data.actual_temps, // From Python
                    borderColor: '#4facfe',
                    backgroundColor: '#4facfe',
                    pointRadius: 6,
                    pointHoverRadius: 9,
                    type: 'scatter' // Show as points
                },
                {
                    label: 'Regression Line (AI Prediction)',
                    data: data.trend_line, // From Python (y_pred)
                    borderColor: '#ff7b54',
                    borderWidth: 3,
                    borderDash: [10, 5],
                    pointRadius: 0,
                    tension: 0 // Straight line
                }
            ]
        },
        options: {
            responsive: true,
            maintainAspectRatio: false,
            interaction: { mode: 'nearest', intersect: true, axis: 'x' },
            plugins: {
                legend: { position: 'top' },
                tooltip: {
                    backgroundColor: 'rgba(0,0,0,0.8)',
                    titleFont: { size: 14 },
                    bodyFont: { size: 14 },
                    padding: 10,
                    cornerRadius: 8,
                    displayColors: false
                }
            },
            scales: {
                y: { display: true, title: { display: true, text: 'Temp (°F)' } },
                x: { grid: { display: false } }
            }
        }
    });
}

// --- 5. SEARCH LISTENER ---
cityInput.addEventListener('keypress', function (e) {
    if (e.key === 'Enter') {
        const city = cityInput.value.trim();
        if(city) getWeather(city);
    }
});